ddl/partition.go

```diff
@@ -1911,27 +1911,152 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 		return ver, errors.Trace(dbterror.ErrPartitionMgmtOnNonpartitioned)
 	}
 
-	newPartitions := make([]model.PartitionDefinition, 0, len(oldIDs))
-	for _, oldID := range oldIDs {
-		for i := 0; i < len(pi.Definitions); i++ {
-			def := &pi.Definitions[i]
-			if def.ID == oldID {
-				pid, err1 := t.GenGlobalID()
-				if err1 != nil {
-					return ver, errors.Trace(err1)
+	switch job.SchemaState {
+	case model.StatePublic:
+		newDefs := make([]model.PartitionDefinition, 0, len(pi.Definitions)-len(oldIDs))
+		pi.OriginalDefinitions = append(pi.OriginalDefinitions, pi.Definitions...)
+		for i := range pi.Definitions {
+			found := false
+			for _, oldID := range oldIDs {
+				if pi.Definitions[i].ID == oldID {
+					found = true
+					break
 				}
-				def.ID = pid
-				// Shallow copy only use the def.ID in event handle.
-				newPartitions = append(newPartitions, *def)
-				break
+			}
+			if !found {
+				newDefs = append(newDefs, pi.Definitions[i])
 			}
 		}
+		pi.Definitions = newDefs
```

Comment:
Would not this mean that the partitions that are being truncated can neither be read or written, i.e. unaccessible and possibly give errors on insert/updates?

Reply:
you are right, the previous logic does not allow concurrent read/writes for truncating partitions. I solve the problem in the new PR.